planner/core/push_down_predicate_to_table_scan.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"sort"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// selectivity = (row count after filter) / (row count before filter), smaller is better
+// income = (1 - selectivity) * restColumnCount, greater is better
+// We use income is avoid the following case:
+// selectivity is 0.7, restColumnCount is 10, income is 3
+// after push down a predicate, selectivity becomes 0.6, restColumnCount becomes 7, income becomes 2.8, which is smaller than 3
+// but the predicate should not be pushed down.
+
+const (
+	selectivityThreshold        = 0.7
+	selectivityImproveThreshold = 0.1
+	// The default now of number of rows in a pack of TiFlash
+	tiflashDataPackSize  = 8192
+	columnCountThreshold = 3
+)
+
+// expressionGroup is used to store
+// 1. a group of expressions
+// 2. the selectivity of the expressions
+type expressionGroup struct {
+	exprs       []expression.Expression
+	selectivity float64
+}
+
+// predicatePushDownToTableScan is used find the selection just above the table scan
+// and try to push down the predicates to the table scan.
+func predicatePushDownToTableScan(sctx sessionctx.Context, plan PhysicalPlan) {
```

Comment:
You can make this function return the result plan:
```
func f(...) PhysicalPlan {
	switch p.(type) {
	case *PhysicalSelection:
		....
		if len(p.Condition) == 0 {
			return p.Children()[0]
		}
	case *TableReader:
		p.tablePlan = f(..)
	default:
		newChildren := make()
		for _, child := range p.Children() {
			append newChildren
		}
		p.SetChildren(newChildren...)
	}
	return p
}
```

Reply:
I think we can avoid the second iteration to remove the empty sel.

Reply:
We need `PhysicalSelection`'s parent.